### ISSUE TITLE
Docs says that no-args constructor of RealmObjects must be empty

### ISSUE THREAD

**Issue by Zhuinden** [realm/realm-java#4941]
https://realm.io/docs/java/latest/#limitations

> If declared, the default constructor (constructor with no parameters) must always be empty. The reason is that a default contructor will call methods which assume a Realm instance is present. But that instance isn’t create before the contructor returns. You can add other constructors for your convienence. 

1.) it should be `convenience`

2.) it should be `isn't created`

3.) you **can** have "default parameters" in the constructor since Realm 2.0 afaik

**Comment by bmeike** [comment]
This bit if text is pretty confusing, top to bottom.  It is impossible to write a "default constructor".  A "default constructor" is the constructor that the compiler supplies, expressly, if the class writer does *not* supply one.  If the writer supplies one, it is by def, not default.

I *think* that all that para means is that you'd better not put anything in your no-args constructor that depends on/changes the state of the object, because anything you do will get over-written by the proxy object is going to change everything anyway.

**Comment by Zhuinden** [comment]
@bmeike there's actually some super-hacky hack going on with that `RealmObjectContext` thing you ran into a few months ago, which allows so that the constructor you define in the class will be used by the `realm.createObject()` call to supply "default parameters" for your RealmObject upon creation.

It has to do with some sort of extreme levels black magic, but it works, except for... something to do with RealmLists? There's an issue for it [here](https://github.com/realm/realm-java/issues/3809) but I don't know the specifics.

**Comment by zaki50** [comment]
@Zhuinden You are right. Setting some values to fields in no-args constructor means giving default values of the fields in `realm.createObject(Class)`. In other cases of instantiating proxy objects, we ignore those values. #3809 is a bug that I couldn't find a good way to fix.

`The reason is that a default contructor will call methods which assume a Realm instance is present. But that instance isn’t create before the contructor returns. ` is  out of date and is the limitation that I removed when I introduced default value support to `realm.createObject(Class)` (and `realm.create*FromJson()`s) in `2.0`.
